unison-cli/src/Unison/Codebase/Editor/HandleInput/Update2.hs

```diff
@@ -128,19 +128,33 @@ handleUpdate2 = do
 
     pure (pped, bigUf)
 
-  -- - typecheck it
-  Cli.respond Output.UpdateStartTypechecking
-  parsingEnv <- makeParsingEnv currentPath namesIncludingLibdeps
-  prettyParseTypecheck bigUf pped parsingEnv >>= \case
-    Left prettyUf -> do
-      Cli.Env {isTranscript} <- ask
-      maybePath <- if isTranscript then pure Nothing else Just . fst <$> Cli.expectLatestFile
-      Cli.respond (Output.DisplayDefinitionsString maybePath prettyUf)
-      Cli.respond Output.UpdateTypecheckingFailure
-    Right tuf -> do
-      Cli.respond Output.UpdateTypecheckingSuccess
-      saveTuf (findCtorNames namesExcludingLibdeps ctorNames Nothing) tuf
-      Cli.respond Output.Success
+  -- If the new-unison-file-to-typecheck is the same as old-unison-file-that-we-already-typechecked, then don't bother
+  -- typechecking again.
+  secondTuf <- do
+    let smallUf = UF.discardTypes tuf
+    let noChanges =
+          and
+            [ UF.dataDeclarations smallUf == UF.dataDeclarations bigUf,
+              UF.effectDeclarations smallUf == UF.effectDeclarations bigUf,
+              UF.terms smallUf == UF.terms bigUf, -- no need to sort these, though it wouldn't hurt
+              UF.watches smallUf == UF.watches bigUf
+            ]
```

Comment:
Seems like you should maybe just have the code that generates `bigUf` return a `Bool` of whether it had to pull in additional dependents.

Also seems like you could just compare sizes here. The `bigUf` can only have additional elements, since you don't remove anything, only add dependents.

Reply:
:+1: I changed this to just compare sizes. Making `bigBuildUnisonFile` return whether it changed anything is a bit more invasive to thread that state through everywhere, so I'm not sure it's worth it.

Reply:
No objections here, just thought I'd mention it.